cpp/src/structure/sort_and_remove_multi_edges_impl.cuh

```diff
@@ -0,0 +1,317 @@
+    thrust::sort(handle.get_thrust_policy(), second_first, pair_first + edgelist_srcs.size());
+  } else {
+    thrust::sort(handle.get_thrust_policy(), pair_first, pair_first + edgelist_srcs.size());
+  }
+
+  edgelist_srcs.resize(
+    thrust::distance(pair_first,
+                     thrust::unique(handle.get_thrust_policy(),
+                                    pair_first,
+                                    pair_first + edgelist_srcs.size(),
+                                    [] __device__(auto lhs, auto rhs) {
+                                      return (thrust::get<0>(lhs) == thrust::get<0>(rhs)) &&
+                                             (thrust::get<1>(lhs) == thrust::get<1>(rhs));
+                                    })),
+    handle.get_stream());
+
+  edgelist_dsts.resize(edgelist_srcs.size(), handle.get_stream());
+
+  return std::make_tuple(std::move(edgelist_srcs), std::move(edgelist_dsts));
+}
+
+template <typename vertex_t, typename edge_value_t>
+std::tuple<rmm::device_uvector<vertex_t>,
+           rmm::device_uvector<vertex_t>,
+           decltype(allocate_dataframe_buffer<edge_value_t>(size_t{0}, rmm::cuda_stream_view{}))>
+sort_and_remove_multi_edges(
+  raft::handle_t const& handle,
+  rmm::device_uvector<vertex_t>&& edgelist_srcs,
+  rmm::device_uvector<vertex_t>&& edgelist_dsts,
+  decltype(allocate_dataframe_buffer<edge_value_t>(0, rmm::cuda_stream_view{}))&& edgelist_values,
+  size_t mem_frugal_threshold)
+{
+  auto pair_first = thrust::make_zip_iterator(edgelist_srcs.begin(), edgelist_dsts.begin());
+
+  if (edgelist_srcs.size() > mem_frugal_threshold) {
+    // Tuning parameter to address high frequency multi-edges
+    size_t num_hash_buckets{16};
+
+    auto hash_counts = compute_hash_sizes(pair_first,
+                                          pair_first + edgelist_srcs.size(),
+                                          num_hash_buckets,
+                                          hash_src_dst_pair<vertex_t>{},
+                                          handle.get_stream());
+
+    auto pivot =
+      static_cast<int32_t>(thrust::distance(hash_counts.begin(),
+                                            thrust::lower_bound(handle.get_thrust_policy(),
+                                                                hash_counts.begin(),
+                                                                hash_counts.end(),
+                                                                edgelist_srcs.size() / 2)));
+
+    auto second_first = detail::mem_frugal_partition(pair_first,
+                                                     pair_first + edgelist_srcs.size(),
+                                                     get_dataframe_buffer_begin(edgelist_values),
+                                                     hash_src_dst_pair<vertex_t>{},
+                                                     pivot,
+                                                     handle.get_stream());
```

Comment:
Same here, you can call https://github.com/rapidsai/cugraph/blob/branch-23.12/cpp/include/cugraph/utilities/shuffle_comm.cuh#L782